src/Data/Tree/Binary.hs

```diff
@@ -1,23 +1,31 @@
-module Data.Tree.Binary (BinaryTree(..), insert, contains)  where
+module Data.Tree.Binary (BinaryTree(..), insert, contains, inOrder)  where
 
 data BinaryTree a
   = Empty
   | Leaf a (BinaryTree a) (BinaryTree a)
   deriving (Eq, Show)
 
 insert :: Ord a => a -> BinaryTree a -> BinaryTree a
-insert newValue leaf =
-  case leaf of
-    Empty -> Leaf newValue Empty Empty
-    Leaf currentValue left right ->
-      if currentValue > newValue then
-        case left of
-          Empty -> Leaf currentValue (Leaf newValue Empty Empty) right
-          _ -> Leaf currentValue (insert newValue left) right
-      else
-        case right of
-          Empty -> Leaf currentValue left (Leaf newValue Empty Empty)
-          _ -> Leaf currentValue left (insert newValue left)
+insert newValue Empty = Leaf newValue Empty Empty
+insert newValue (Leaf currentValue left right)
+  | currentValue > newValue =
+    case left of
+      Empty -> Leaf currentValue (Leaf newValue Empty Empty) right
+      _ -> Leaf currentValue (insert newValue left) right
```

Comment:
So, just to be a pain in the ass, I'll note that a duplication remains here that can be reduced by recursion -- note the equivalent occurences of `Leaf newValue Empty Empty`.